### PR TITLE
Work on virtuals, fix liquids crashing the game when falling outside of a level and other stuff

### PR DESCRIPTION
**Pull request by Mr-Auto**

### CODE REVIEW

File: src/game_api/entities_items.hpp [hunk 1]
```diff
@@ -65,14 +75,24 @@ class VladsCape : public Cape
     bool can_double_jump;
 };
 
-class Mattock : public Movable
+class Purchesable : public Movable
```
Comment (gmjosack) [inline src/game_api/entities_items.hpp:78]: Should this be spelled `Purchasable`?

I keep reading it as per-cheese-able

File: src/game_api/script/usertypes/entities_items_lua.cpp [hunk 2]
```diff
@@ -10,6 +10,10 @@ namespace NEntitiesItems
 void register_usertypes(sol::state& lua)
 {
     lua["Entity"]["as_bomb"] = &Entity::as<Bomb>;
+    lua["Entity"]["as_projectile"] = &Entity::as<Projectile>;
+    lua["Entity"]["as_Purchasable"] = &Entity::as<Purchasable>;
+    lua["Entity"]["as_bow"] = &Entity::as<Bow>;
+    lua["Entity"]["as_dummyPurchasableentity"] = &Entity::as<DummyPurchasableEntity>;
```
Comment (gmjosack) [inline src/game_api/script/usertypes/entities_items_lua.cpp:16]: any reason these are capitalized? no other methods have capital case.

File: docs/script-api.md [hunk 3]
```diff
@@ -5060,6 +5101,7 @@ For reference, the available `as_<typename>` functions are listed below:
 - as_dmspawning
 - as_door
 - as_drill
+- as_dummyPurchasableentity
```
Comment (gmjosack) [inline docs/script-api.md:5104]: looks like docs need to be regenerated

File: src/game_api/entities_items.hpp [hunk 4]
```diff
@@ -65,14 +75,24 @@ class VladsCape : public Cape
...
+class Mattock : public Purchasable
 {
   public:
     uint8_t remaining;
 };
 
-class Gun : public Movable
-{
+class Gun : public Purchasable
+{ // BIG NOTE: scepter is not a Purchasable, but that's the only Gun that isn't so i din't bother to make new type for it
```
Comment (gmjosack) [inline src/game_api/entities_items.hpp:95]: Wouldn't this expose a non existent `buy` virtual for the scepter then? I imagine anyone calling that would be undefined and bad?

Comment (Mr-Auto) [reply src/game_api/entities_items.hpp:95]: only stuff in src/game_api/script is exposed to lua, and also we decide what to expose, it's not automatic

until i know this function can be used in some way or another, there is no reason to add another pointless type for one entity, i added the rest only cuz it was relatively easy

File: src/game_api/level_api.hpp [hunk 5]
```diff
@@ -1,6 +1,7 @@
 #pragma once
 
 #include "game_allocator.hpp"
+#include "imgui.h"
```
Comment (Malacath-92) [inline src/game_api/level_api.hpp:4]: Why is this here?

File: src/game_api/level_api.hpp [hunk 6]
```diff
@@ -458,9 +460,17 @@ struct LevelGenSystem
     std::uint32_t spawn_room_y;
     float spawn_x;
     float spawn_y;
-    DoorCoords* exit_doors_locations;
-    void* unknown37;
-    void* unknown38;
+    union
+    {
+        std::vector<ImVec2> exit_doors;
```
Comment (Malacath-92) [inline src/game_api/level_api.hpp:465]: Oh, I see it's for this. I'd prefer if we don't depend on imgui on the game types, that will make it more annoying later to split just the types and functionality (for pulling out types into their own headers as optional additions to the .dll interface).

Comment (Mr-Auto) [reply src/game_api/level_api.hpp:465]: I used this just cuz we already have this in API, exposed etc.

Comment (Malacath-92) [reply src/game_api/level_api.hpp:465]: Well I will have to remove it later either way.

Comment (Mr-Auto) [reply src/game_api/level_api.hpp:465]: ok, then i will make some quick XY struct

Comment (Malacath-92) [reply src/game_api/level_api.hpp:465]: Works for me  :)
And sorry, I didn't mean to be rude just there. It made perfect sense what you did without the context.

File: src/game_api/script/usertypes/entities_activefloors_lua.cpp [hunk 7]
```diff
@@ -120,6 +118,8 @@ void register_usertypes(sol::state& lua)
         sol::base_classes,
         sol::bases<Entity, Movable>());
 
+    falling_platform["emitted_light"] = &FallingPlatform::timer; // wrong name, kept for backwards compatibility
```
Comment (Dregu) [inline src/game_api/script/usertypes/entities_activefloors_lua.cpp:121]: When a mistake is this obvious, I think it's the users fault for using it and not saying anything, so we can break backwards compatibility.

File: src/game_api/entity.cpp [hunk 8]
```diff
@@ -716,3 +716,16 @@ void Entity::set_pre_collision2(std::uint32_t reserved_callback_id, std::functio
...
+{
+    if (items.size)
+        return std::span<uint32_t>(items.uids().begin(), items.uids().end());
+
+    return {};
+}
+
+void Movable::light_on_fire_broken()
+{
+    this->light_on_fire(0x64);
```
Comment (Malacath-92) [inline src/game_api/entity.cpp:730]: Why exactly `0x64`, maybe put the answer in a comment?

File: src/game_api/level_api.hpp [hunk 9]
```diff
@@ -499,6 +508,8 @@ struct LevelGenSystem
 
     uint32_t get_procedural_spawn_chance(uint32_t chance_id);
     bool set_procedural_spawn_chance(uint32_t chance_id, uint32_t inverse_chance);
+
+    ~LevelGenSystem(){}; // cuz it was complaining
```
Comment (Malacath-92) [inline src/game_api/level_api.hpp:512]: Where was it complaining? Are you invoking this function somewhere? Otherwise it would not complain, that needs to be fixed.

File: src/game_api/math.hpp [hunk 10]
```diff
@@ -89,12 +89,114 @@ struct AABB
     float bottom{0};
 };
 
+struct XY
```
Comment (Malacath-92) [inline src/game_api/math.hpp:92]: This should probably be `Vec2` or something like that. No one in the history of linear algebra code called this `XY`  😜
If you want to make sure it doesn't clash with anything else put it into a `Spel2` namespave or something along those lines.

File: src/game_api/script/usertypes/entities_monsters_lua.cpp [hunk 11]
```diff
@@ -269,6 +269,8 @@ void register_usertypes(sol::state& lua)
         &Shopkeeper::name,
         "shotgun_attack_delay",
         &Shopkeeper::shotgun_attack_delay,
+        "have_key",
```
Comment (Malacath-92) [inline src/game_api/script/usertypes/entities_monsters_lua.cpp:272]: Should this not be `has_key`?

File: src/game_api/entity.cpp (hunk 8 above)
Comment (Mr-Auto) [reply src/game_api/entity.cpp:730]: The game normally lights you up for 100 frames, this is hard coded value.

I guess there could be some sources that set it to different times, but. This is a broken version of this function that shouldn't be used, kept for backwards compatibility. It doesn't show in the doc

File: src/game_api/level_api.hpp (hunk 9 above)
Comment (Mr-Auto) [reply src/game_api/level_api.hpp:512]: When i added the `union` compiler started throwing `warning C4624: 'LevelGenSystem': destructor was implicitly defined as deleted`, i don't get why in this scenario it cares about the deconstructor, unless there is some code that calls it

File: src/game_api/entity.cpp (hunk 8 above)
Comment (Malacath-92) [reply src/game_api/entity.cpp:730]: Sorry for the confusion, I got why. My point was just that it's better to put a comment for future us when we're not currently looking at a PR. Magic numbers are not cool.

File: src/game_api/level_api.hpp (hunk 9 above)
Comment (Malacath-92) [reply src/game_api/level_api.hpp:512]: It does sound like there is code that calls it, otherwise it should complain. If you look for the error in the Output window there should be more info.

Comment (Mr-Auto) [reply src/game_api/level_api.hpp:512]: There is nothing more the in the output, just the standard `the following warning is treated as an error` and then the warning that i posted on `level_api.hpp(511)`

Before that it builds the render_api which has only one line ralated to the level_api and it's `texture_id = State::get().ptr_local()->current_theme->get_dynamic_texture(texture_id);`

File: src/game_api/math.hpp (hunk 10 above)
Comment (Mr-Auto) [reply src/game_api/math.hpp:92]: where does the '2' come from?

Comment (Malacath-92) [reply src/game_api/math.hpp:92]: It has two elements. As opposed to a `Vec3` which has 3 elements, etc. Same goes for matrices that you'd call `Mat2` and not `XXXYYXYY`

Comment (Mr-Auto) [reply src/game_api/math.hpp:92]: btw. i was thinking about constructor and split function from AABB, but not sure which corners i should choose :thinking:

**Comment by Mr-Auto** [comment]
btw. i don't know how to fix the problem with generate.py :(

**Comment by Malacath-92** [comment]
I'm gonna pull the PR and check for the two issues.
